airflow/jobs.py

```diff
@@ -1918,119 +1971,131 @@ def _manage_executor_state(self, started):
+        :type dag_run: models.DagRun
+        """
         tasks_to_run = {}
-        failed = set()
-        succeeded = set()
-        started = {}
-        skipped = set()
-        not_ready = set()
-        deadlocked = set()
-
-        # create dag runs
-        dr_start_date = start_date or min([t.start_date for t in self.dag.tasks])
-        end_date = end_date or datetime.now()
-        # next run date for a subdag isn't relevant (schedule_interval for subdags
-        # is ignored) so we use the dag run's start date in the case of a subdag
-        next_run_date = (self.dag.normalize_schedule(dr_start_date)
-                         if not self.dag.is_subdag else dr_start_date)
 
-        active_dag_runs = []
-        while next_run_date and next_run_date <= end_date:
-            run_id = BackfillJob.ID_FORMAT_PREFIX.format(next_run_date.isoformat())
-
-            # check if we are scheduling on top of a already existing dag_run
-            # we could find a "scheduled" run instead of a "backfill"
-            run = DagRun.find(dag_id=self.dag.dag_id,
-                              execution_date=next_run_date,
-                              session=session)
-            if not run:
-                run = self.dag.create_dagrun(
-                    run_id=run_id,
-                    execution_date=next_run_date,
-                    start_date=datetime.now(),
-                    state=State.RUNNING,
-                    external_trigger=False,
-                    session=session,
-                )
-            else:
-                run = run[0]
-
-            # set required transient field
-            run.dag = self.dag
-
-            # explictely mark running as we can fill gaps
-            run.state = State.RUNNING
-            run.run_id = run_id
-            run.verify_integrity(session=session)
+        if not dag_run:
+            return tasks_to_run
 
-            self.reset_state_for_orphaned_tasks(filter_by_dag_run=run, session=session)
+        # check if we have orphaned tasks
+        self.reset_state_for_orphaned_tasks(filter_by_dag_run=dag_run, session=session)
 
-            # for some reason if we dont refresh the reference to run is lost
-            run.refresh_from_db()
-            make_transient(run)
-            active_dag_runs.append(run)
+        # for some reason if we don't refresh the reference to run is lost
+        dag_run.refresh_from_db()
+        make_transient(dag_run)
 
-            for ti in run.get_task_instances():
-                # all tasks part of the backfill are scheduled to run
-                if ti.state == State.NONE:
-                    ti.set_state(State.SCHEDULED, session=session)
-                tasks_to_run[ti.key] = ti
+        for ti in dag_run.get_task_instances():
+            # all tasks part of the backfill are scheduled to run
+            if ti.state == State.NONE:
+                ti.set_state(State.SCHEDULED, session=session)
```

Comment:
Right, as of this PR I just organized the code to have it easier to make those changes, so this is just a re-org of the code. I could create a ticket to improve the performance here and make a follow up change unless someone else thinks it should be included on this PR @aoen @bolkedebruin